packages/eslint-plugin/rules/no-top-level-material-ui-4-imports.js

```diff
@@ -0,0 +1,171 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-check
+
+const KNOWN_STYLES = [
+  'makeStyles',
+  'withStyles',
+  'createStyles',
+  'styled',
+  'useTheme',
+  'Theme',
+];
+
+/** @type {import('eslint').Rule.RuleModule} */
+module.exports = {
+  meta: {
+    type: 'problem',
+    fixable: 'code',
+    messages: {
+      topLevelImport: 'Top level imports for Material UI are not allowed',
+    },
+    docs: {
+      description: 'Forbid top level import from Material UI v4 packages.',
+      url: 'https://github.com/backstage/backstage/blob/master/packages/eslint-plugin/docs/rules/no-top-level-material-ui-4-imports.md',
+    },
+  },
+  create: context => ({
+    ImportDeclaration: node => {
+      // Anatomy of a Node
+      // Example: import SvgIcon, { SvgIconProps } from '@material-ui/core/SvgIcon';
+      // Specifiers are the part between the `import` and `from`, in the example that would be `SvgIcon, { SvgIconProps }`
+      // Source is the part after the `from`, in the example that would be `'@material-ui/core/SvgIcon'`
+      // Source value gets you `@material-ui/core/SvgIcon` without the quotes, where as Source raw gets it as is
+
+      // Return if empty import
+      if (node.specifiers.length === 0) return;
+      // Return if empty source value
+      if (!node.source.value) return;
+      // Return if source value not a string
+      if (typeof node.source.value !== 'string') return;
+      // Return if import does not start with '@material-ui/'
+      if (!node.source.value.startsWith('@material-ui/')) return;
+      // Return if import is from '@material-ui/core/styles', as it's valid already
+      if (node.source.value === '@material-ui/core/styles') return;
+      // Return if proper import eg. `import Box from '@material-ui/core/Box'`
+      if (
+        node.specifiers.length >= 1 &&
+        node.source.value?.split('/').length === 3
+      )
+        return;
+
+      // Return if third level or deeper imports
+      if (
+        node.specifiers.length >= 1 &&
+        node.source.value.split('/').length > 3
+      ) {
+        return;
+      }
+
+      // Report all other imports
+      context.report({
+        node,
+        messageId: 'topLevelImport',
+        fix: fixer => {
+          const replacements = [];
+          const styles = [];
+
+          const specifiers = node.specifiers.filter(
+            s => s.type === 'ImportSpecifier',
+          );
+
+          const specifiersMap = specifiers.map(s => {
+            const propsTest = /^([A-Z]\w+)Props$/.test(s.local.name);
```

Comment:
to not have to repeat the regex twice, you can just do the `exec` below first, and do `const propsTest = propsMatch !== null` instead.

Or skip this variable entirely, and just check the exec value where needed below - you could just as well do `propValue: propsMatch ? propsMatch[1] ? undefined` for example, or even `propValue: propsMatch?.[1]` if you wanna get fancy

Reply:
Sweet, did not like the double regex, this works nicely, thanks! 👍